app/repositories/ingame_logins.py

```diff
@@ -0,0 +1,233 @@
+# +--------------+------------------------+------+-----+---------+-------+
+
+READ_PARAMS = textwrap.dedent(
+    """\
+        id, userid, ip, osu_ver, osu_stream, datetime
+    """,
+)
+
+
+class IngameLogin(TypedDict):
+    id: int
+    userid: str
+    ip: str
+    osu_ver: str
+    osu_stream: str
+    datetime: datetime
+
+
+class InGameLoginUpdateFields(TypedDict, total=False):
+    userid: str
+    ip: str
+    osu_ver: str
+    osu_stream: str
+
+
+async def create(
+    user_id: int,
+    ip: str,
+    osu_ver: str,
+    osu_stream: str,
+) -> IngameLogin:
+    """Create a new login entry in the database."""
+    query = f"""\
+        INSERT INTO ingame_logins (userid, ip, osu_ver, osu_stream, datetime)
+             VALUES (:userid, :ip, :osu_ver, :osu_stream, NOW())
+    """
+    params = {
+        "userid": user_id,
+        "ip": ip,
+        "osu_ver": osu_ver,
+        "osu_stream": osu_stream,
+    }
+    rec_id = await app.state.services.database.execute(query, params)
+
+    query = f"""\
+        SELECT {READ_PARAMS}
+          FROM ingame_logins
+         WHERE id = :id
+    """
+    params = {
+        "id": rec_id,
+    }
+    ingame_login = await app.state.services.database.fetch_one(query, params)
+
+    assert ingame_login is not None
+    return cast(IngameLogin, ingame_login)
+
+
+async def fetch_one(
+    id: int | None = None,
+    user_id: int | None = None,
+    ip: str | None = None,
+    osu_ver: str | None = None,
+    osu_stream: str | None = None,
+) -> IngameLogin | None:
+    """Fetch a login entry from the database."""
+    if (
+        id is None
+        and user_id is None
+        and ip is None
+        and osu_ver is None
+        and osu_stream is None
+    ):
+        raise ValueError("Must provide at least one parameter.")
+
+    query = f"""\
+        SELECT {READ_PARAMS}
+          FROM ingame_logins
+         WHERE id = COALESCE(:id, id)
+           AND userid = COALESCE(:userid, userid)
+           AND ip = COALESCE(:ip, ip)
+           AND osu_ver = COALESCE(:osu_ver, osu_ver)
+           AND osu_stream = COALESCE(:osu_stream, osu_stream)
+    """
+    params = {
+        "id": id,
+        "userid": user_id,
+        "ip": ip,
+        "osu_ver": osu_ver,
+        "osu_stream": osu_stream,
+    }
+    ingame_login = await app.state.services.database.fetch_one(query, params)
+
+    return cast(IngameLogin, ingame_login) if ingame_login is not None else None
+
+
+async def fetch_count(
+    user_id: int | None = None,
+    ip: str | None = None,
+) -> int:
+    """Fetch the number of logins in the database."""
+    query = """\
+        SELECT COUNT(*) AS count
+          FROM ingame_logins
+        WHERE userid = COALESCE(:userid, userid)
+          AND ip = COALESCE(:ip, ip)
+    """
+    params = {
+        "userid": user_id,
+        "ip": ip,
+    }
+    rec = await app.state.services.database.fetch_one(query, params)
+    assert rec is not None
+    return rec["count"]
```

Comment:
rec._mapping["count"] is fine here - no need to make a copy because we do not need to return a mutable copy